nixos/modules/services/matrix/synapse/workers.nix

```
@@ -0,0 +1,180 @@
+              default = "synapse.app.generic_worker";
+            };
+            worker_listeners = mkOption {
+              type = types.listOf workerListenerType;
+              default = [ ];
+              description = "Listener config for the worker, similar to the main synapse listener config";
+            };
+          };
+        };
+        default = { };
+        description = "A freeform option of the workers settings, see upstream synapse documentation for options";
+      };
+    });
+
+    workerListenerType = types.submodule {
+      options = {
+        type = mkOption {
+          type = types.enum [ "http" "metrics" ];
+          description = "The type of the listener";
+          default = "http";
+        };
+        port = mkOption {
+          type = types.port;
+          description = "The TCP port to bind to";
+        };
+        bind_addresses = mkOption {
+          type = with types; listOf str;
+          description = "A list of local addresses to listen on";
+          default = [ "127.0.0.1" ];
+        };
+        tls = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to enable TLS for this listener.
+            Will use the TLS key/cert specified in tls_private_key_path / tls_certificate_path.
+          '';
+          default = false;
+        };
+        x_forwarded = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to use the X-Forwarded-For HTTP header as the client IP.
+            This option is only valid for an 'http' listener.
+            It is useful when Synapse is running behind a reverse-proxy.
+          '';
+          default = true;
+        };
+        resources = mkOption {
+          type = types.listOf (types.submodule {
+            options = {
+              names = mkOption {
+                type = types.listOf (types.enum [
+                  "client"
+                  "consent"
+                  "federation"
+                  "health"
+                  "keys"
+                  "media"
+                  "metrics"
+                  "openid"
+                  "replication"
+                  "static"
+                ]);
+                description = "A list of resources to host on this port";
+                default = [ ];
+                example = lib.literalExpression ''[ "client" ]'';
+              };
+              compress = lib.mkEnableOption "HTTP compression for this resource";
+            };
+          });
+          default = [ { } ];
+          description = "The resources to make available on this listener";
+        };
+      };
+    };
+
+  in {
+    instances = mkOption {
+      type = types.attrsOf workerInstanceType;
+      default = { };
+      example = {
+        "federation_sender1" = {
+          settings = {
+            worker_name = "federation_sender1";
+            worker_app = "synapse.app.generic_worker";
+
```

```suggestion
        "federation_sender1" = {
          settings = {
            worker_name = "federation_sender1";
```
we can make the example a bit more on point by removing the default value and it would be cool if we could infer worker_name from the attr name

---

We do actually, so we can just remove it. I'll show off a more advanced worker rather than the super simple federation_sender